perun-core/src/main/java/cz/metacentrum/perun/core/impl/UsersManagerImpl.java

```diff
@@ -1146,6 +1157,99 @@ public boolean isLoginReserved(PerunSession sess, String namespace, String login
 		}
 	}
 
+	@Override
+	public List<Pair<String, String>> getAllBlockedLoginsInNamespaces(PerunSession sess) {
+		try {
+			return jdbc.query("select login, namespace from blocked_logins", BLOCKED_LOGINS_MAPPER);
+		} catch (RuntimeException e) {
+			throw new InternalErrorException(e);
+		}
+	}
+
+	@Override
+	public boolean isLoginBlocked(PerunSession sess, String login) {
+		Utils.notNull(login, "userLogin");
+
+		try {
+			return jdbc.queryForInt("select count(1) from blocked_logins where login=?", login) == 1;
+		} catch (RuntimeException e) {
+			throw new InternalErrorException(e);
+		}
+	}
+
+	@Override
+	public boolean isLoginBlockedGlobally(PerunSession sess, String login) {
+		Utils.notNull(login, "userLogin");
+
+		try {
+			return jdbc.queryForInt("select count(1) from blocked_logins where login=? and namespace is null", login) == 1;
+		} catch (RuntimeException e) {
+			throw new InternalErrorException(e);
+		}
+	}
+
+	@Override
+	public boolean isLoginBlockedForNamespace(PerunSession sess, String login, String namespace) {
+		Utils.notNull(login, "userLogin");
+
+		try {
+			if (namespace == null) {
+				return this.isLoginBlockedGlobally(sess, login);
+			} else {
+				return jdbc.queryForInt("select count(1) from blocked_logins where login=? and namespace=?", login, namespace) == 1;
+			}
+		} catch (RuntimeException e) {
+			throw new InternalErrorException(e);
+		}
+	}
+
+	@Override
+	public void blockLogin(PerunSession sess, String login, String namespace) throws LoginIsAlreadyBlockedException {
+		Utils.notNull(login, "userLogin");
+
+		try {
+			int newId = Utils.getNewId(jdbc, "blocked_logins_id_seq");
+			jdbc.update("insert into blocked_logins(id, login, namespace) values (?,?,?)",
+				newId, login, namespace);
+			if (namespace == null) {
+				log.info("Login {} globally blocked", login);
+			} else {
+				log.info("Login {} blocked in namespace: {}", login, namespace);
+			}
+		} catch (DuplicateKeyException ex) {
+			if (namespace == null) {
+				throw new LoginIsAlreadyBlockedException("Login: " + login + " is already blocked globally");
+			} else {
+				throw new LoginIsAlreadyBlockedException("Login: " + login + " is already blocked in namespace: " + namespace);
+			}
+		} catch (RuntimeException ex) {
+			throw new InternalErrorException(ex);
+		}
+	}
+
+	@Override
+	public void unblockLogin(PerunSession sess, String login, String namespace) throws LoginIsNotBlockedException {
+		Utils.notNull(login, "userLogin");
+
+		try {
+			if (namespace == null) {
+				int numAffected = jdbc.update("delete from blocked_logins where login=? and namespace is null", login);
```

Comment:
```suggestion
				int numAffected = jdbc.update("delete from blocked_logins where login=? and namespace " + namespace == null ? "is null" : "=?", login);
```
The logging makes it bit more awkward now that you mention it, but then you could just leave the log calls in the if-else, which would still be less duplicate code.
It's not big of a deal, go with what seems better readable to you